File: go/core/llm_provider.go
package core

import (
	"context"
	"errors"
)

// LlmProvider is the LLM-call seam the agent loop depends on. It is the existing
// ChatClient interface under a name that names the role — formalizing the seam so
// the agent loop is unit-testable deterministically, without a live model or
// network. The GatewayClient already implements it; tests inject a MockLlmProvider.
//
// This keeps backward compatibility: NewSmoothAgent still takes a ChatClient, and
// any LlmProvider is a ChatClient (and vice versa) since they are the same shape.
type LlmProvider = ChatClient

// WithUsage returns a copy of resp with the given token usage attached. Handy when
// scripting the mock so the streaming final chunk / non-streaming Usage is non-zero.
func WithUsage(resp ChatResponse, promptTokens, completionTokens int) ChatResponse {
	resp.Usage = Usage{PromptTokens: promptTokens, CompletionTokens: completionTokens}
	return resp
}

// ScriptedUsage is the token usage a SCRIPTED mock response reports. Fixed, and
// identical in all five engines' mocks (Rust · Go · Python · TypeScript · C#), so
// the shared server scenario corpus can assert eventual_response.usage as a real
// cross-language invariant instead of documenting five different answers (pearl
// th-4f1263).
//
// Only the FIFO scripting helpers (PushText / PushToolCall) attach it. An
// unscripted response — the benign empty reply a drained script falls back to —
// still reports nothing, so "the script ran out" stays distinguishable from "the
// model answered". Pass an explicit WithUsage to PushResponse to override.
func ScriptedUsage() Usage {
	return Usage{PromptTokens: 10, CompletionTokens: 5}
}

// TextResponse builds a plain-text ChatResponse (no tool calls). Handy for
// scripting the mock and for assertions.
func TextResponse(content string) ChatResponse {
	return ChatResponse{Content: content}
}

// ToolCallResponse builds a ChatResponse that requests a single tool call.
// arguments is the raw JSON-string the model emits for the call's arguments.
func ToolCallResponse(id, name, arguments string) ChatResponse {
	return ChatResponse{ToolCalls: []ToolCall{{ID: id, Name: name, Arguments: arguments}}}
}

// scriptedOutcome is one entry in the mock's script: a response or an error.
type scriptedOutcome struct {
	resp ChatResponse
	err  error
}

// MockLlmProvider is a deterministic LlmProvider for tests. Script the responses
// it should return (FIFO), drive your code, then assert on Calls. Build it up
// fluently with PushText / PushToolCall / PushError.
//
// It replaces the ad-hoc fakeClient the tests rolled by hand, and mirrors the
// BEHAVIOR of the Rust reference's MockLlmClient (record + replay). It is not
// safe for concurrent use — a turn drives it serially, which is the intended use.
type MockLlmProvider struct {
	script []scriptedOutcome
	calls  []ChatRequest
}

// NewMockLlmProvider returns an empty mock. Script it with the Push* methods.
func NewMockLlmProvider() *MockLlmProvider {
	return &MockLlmProvider{}
}

// PushResponse queues a raw ChatResponse for the next Chat call.
func (m *MockLlmProvider) PushResponse(resp ChatResponse) *MockLlmProvider {
	m.script = append(m.script, scriptedOutcome{resp: resp})
	return m
}

// PushText queues a plain-text response for the next Chat call, reporting
// ScriptedUsage.
func (m *MockLlmProvider) PushText(content string) *MockLlmProvider {
	resp := TextResponse(content)
	resp.Usage = ScriptedUsage()
	return m.PushResponse(resp)
}

// PushToolCall queues a single-tool-call response for the next Chat call,
// reporting ScriptedUsage.
func (m *MockLlmProvider) PushToolCall(id, name, arguments string) *MockLlmProvider {
	resp := ToolCallResponse(id, name, arguments)
	resp.Usage = ScriptedUsage()
	return m.PushResponse(resp)
}

// PushError queues an error to be returned from the next Chat call.
func (m *MockLlmProvider) PushError(message string) *MockLlmProvider {
	m.script = append(m.script, scriptedOutcome{err: errors.New(message)})
	return m
}

// Calls returns every request the mock has received so far, in order.
func (m *MockLlmProvider) Calls() []ChatRequest {
	return m.calls
}

// CallCount returns the number of requests received.
func (m *MockLlmProvider) CallCount() int {
	return len(m.calls)
}

// LastCall returns the most recent request and true, or a zero request and false
// if none have been received.
func (m *MockLlmProvider) LastCall() (ChatRequest, bool) {
	if len(m.calls) == 0 {
		return ChatRequest{}, false
	}
	return m.calls[len(m.calls)-1], true
}

// Chat implements ChatClient / LlmProvider: it records the request, then replays
// the next scripted outcome. With an empty script it returns a benign empty text
// response so loops terminate cleanly.
func (m *MockLlmProvider) Chat(_ context.Context, req ChatRequest) (ChatResponse, error) {
	m.calls = append(m.calls, req)
	next, ok := m.pop()
	if !ok {
		return ChatResponse{}, nil
	}
	if next.err != nil {
		return ChatResponse{}, next.err
	}
	return next.resp, nil
}

// ChatStream implements StreamingChatClient: it records the request, then replays
// the SAME next scripted outcome as Chat, but as chunked deltas. Text is split into
// a few content-delta chunks; each tool call is split into an opening chunk (ID +
// name + first half of arguments) and a second chunk with the rest of the arguments
// (exercising the agent's index-keyed accumulator); a final chunk carries usage. A
// scripted error is returned synchronously (connect-time), mirroring a failed open.
func (m *MockLlmProvider) ChatStream(_ context.Context, req ChatRequest) (<-chan ChatChunk, error) {
	m.calls = append(m.calls, req)
	next, ok := m.pop()
	if !ok {
		next = scriptedOutcome{resp: TextResponse("")}
	}
	if next.err != nil {
		return nil, next.err
	}
	ch := make(chan ChatChunk)
	resp := next.resp
	go func() {
		defer close(ch)
		for _, piece := range splitIntoChunks(resp.Content, 3) {
			ch <- ChatChunk{ContentDelta: piece}
		}
		for i, tc := range resp.ToolCalls {
			// Rune split, same reason as splitIntoChunks: arguments are JSON and a
			// value like {"city":"München"} splits mid-rune on a byte midpoint.
			args := []rune(tc.Arguments)
			mid := len(args) / 2
			ch <- ChatChunk{ToolCallDeltas: []ToolCallDelta{{Index: i, ID: tc.ID, Name: tc.Name, ArgsFragment: string(args[:mid])}}}
			ch <- ChatChunk{ToolCallDeltas: []ToolCallDelta{{Index: i, ArgsFragment: string(args[mid:])}}}
		}
		u := resp.Usage
		ch <- ChatChunk{Usage: &u}
	}()
	return ch, nil
}

// pop removes and returns the next scripted outcome, or false if the script is empty.
func (m *MockLlmProvider) pop() (scriptedOutcome, bool) {
	if len(m.script) == 0 {
		return scriptedOutcome{}, false
	}
	next := m.script[0]
	m.script = m.script[1:]
	return next, true
}

// splitIntoChunks splits s into up to n roughly-equal non-empty pieces.
//
// Splits on RUNE boundaries, not bytes: a byte split lands mid-rune on any
// multi-byte character (an em-dash is 3 bytes, an emoji 4) and each fragment is
// then invalid UTF-8, which json.Marshal replaces with U+FFFD — permanently, so
// reassembling the chunks no longer recovers the text.
func splitIntoChunks(s string, n int) []string {
	r := []rune(s)
	if len(r) == 0 {
		return nil
	}
	parts := n
	if len(r) < parts {
		parts = len(r)
	}
	if parts < 1 {
		parts = 1
	}
	size := (len(r) + parts - 1) / parts // ceil
	var out []string
	for i := 0; i < len(r); i += size {
		end := i + size
		if end > len(r) {
			end = len(r)
		}
		out = append(out, string(r[i:end]))
	}
	return out
}

File: typescript/core/src/llmProvider.ts
/**
 * An `LlmProvider` seam over the LLM call so the agentic loop can be unit-tested
 * deterministically, without a live model or network.
 *
 * The agent already takes an injected OpenAI-compatible chat client
 * ({@link ChatClientLike}). This module *formalizes* that as the provider seam —
 * `LlmProvider` is an alias of `ChatClientLike`, so the existing `SmoothAgent`
 * constructor is unchanged and backward compatible (the real `openai` SDK still
 * satisfies it).
 *
 * It also ships a reusable, exported {@link MockLlmProvider} that replaces the
 * ad-hoc fake clients the tests rolled by hand. The mock:
 *
 * - is constructed with a script of responses — plain text, tool-call responses,
 *   and errors;
 * - returns them in FIFO order across calls;
 * - records each request (the messages + tool specs it was given) so a test can
 *   assert on what the agent sent.
 *
 * This mirrors the BEHAVIOR of the Rust reference's `MockLlmClient`
 * (`rust/smooth-operator-core/src/llm_provider.rs`). The mock implements both the
 * non-streaming `create` seam (used by {@link SmoothAgent.run}) and the streaming
 * `createStream` seam (used by {@link SmoothAgent.runStream}): it replays the SAME
 * FIFO script as chunked deltas — text split into a few pieces, tool-call
 * `arguments` split across two chunks to exercise the accumulator, and a final
 * chunk carrying usage. Structured-output lands when that feature lands here.
 */

import type { ChatChunk, ChatClientLike } from './agent.js';

/** The LLM call surface the agent loop depends on. Identical to {@link ChatClientLike}. */
export type LlmProvider = ChatClientLike;

/** An OpenAI-shaped assistant message — the `choices[0].message` the agent reads. */
export interface ScriptedMessage {
    content: string | null;
    tool_calls?: Array<{ id: string; function: { name: string; arguments: string } }> | null;
}

/** Build a plain-text scripted response (no tool calls). */
export function textResponse(content: string): ScriptedMessage {
    return { content };
}

/** Build a scripted response that requests a single tool call. */
export function toolCallResponse(id: string, name: string, args: string): ScriptedMessage {
    return { content: null, tool_calls: [{ id, function: { name, arguments: args } }] };
}

/** One request the mock received, captured for assertions. */
export interface RecordedCall {
    /** The full request body passed to `chat.completions.create`. */
    body: Record<string, unknown>;
    /** The messages passed on this call. */
    messages: Array<Record<string, unknown>>;
    /** The tool specs offered to the model, if any. */
    tools?: Array<Record<string, unknown>>;
}

/** Optional token usage to attach to a scripted response (the model/gateway reports it). */
export interface ScriptedUsage {
    prompt_tokens?: number;
    completion_tokens?: number;
}

/** A scripted outcome: either a response message (with optional usage) or an error to throw. */
type Outcome = { kind: 'message'; message: ScriptedMessage; usage?: ScriptedUsage } | { kind: 'error'; message: string };

/**
 * A deterministic {@link LlmProvider} for tests. Script the responses it should
 * return (FIFO), drive your code, then assert on {@link MockLlmProvider.calls}.
 *
 * Construct empty and build up fluently (`pushText` / `pushToolCall` / `pushError`),
 * or pass an initial script of {@link ScriptedMessage}s.
 *
 * @example
 * const mock = new MockLlmProvider();
 * mock.pushText('hello there');
 * const agent = new SmoothAgent(mock, {});
 * const result = await agent.run('hi');
 * expect(result.text).toBe('hello there');
 * expect(mock.callCount).toBe(1);
 */
export class MockLlmProvider implements ChatClientLike {
    private readonly script: Outcome[] = [];
    private readonly recorded: RecordedCall[] = [];

    constructor(script: ScriptedMessage[] = []) {
        this.script = script.map((message) => ({ kind: 'message', message }));
    }

    // ── scripting (fluent: each returns this) ────────────────────────────────

    /** Queue a raw OpenAI-shaped assistant message (with optional usage) for the next call. */
    pushResponse(message: ScriptedMessage, usage?: ScriptedUsage): this {
        this.script.push({ kind: 'message', message, usage });
        return this;
    }

    /** Queue a plain-text response (with optional usage) for the next call. */
    pushText(content: string, usage?: ScriptedUsage): this {
        return this.pushResponse(textResponse(content), usage);
    }

    /** Queue a single-tool-call response (with optional usage) for the next call. */
    pushToolCall(id: string, name: string, args: string, usage?: ScriptedUsage): this {
        return this.pushResponse(toolCallResponse(id, name, args), usage);
    }

    /** Queue an error to be thrown on the next call. */
    pushError(message: string): this {
        this.script.push({ kind: 'error', message });
        return this;
    }

    // ── recordings ───────────────────────────────────────────────────────────

    /** Every request the mock has received so far, in order. */
    get calls(): readonly RecordedCall[] {
        return this.recorded;
    }

    /** Number of requests received. */
    get callCount(): number {
        return this.recorded.length;
    }

    /** The most recent request, if any. */
    get lastCall(): RecordedCall | undefined {
        return this.recorded[this.recorded.length - 1];
    }

    // ── the ChatClientLike surface ───────────────────────────────────────────

    private record(body: Record<string, unknown>): void {
        this.recorded.push({
            body,
            messages: (body.messages as Array<Record<string, unknown>>) ?? [],
            tools: body.tools as Array<Record<string, unknown>> | undefined,
        });
    }

    readonly chat = {
        completions: {
            create: async (body: Record<string, unknown>) => {
                this.record(body);
                const outcome = this.script.shift();
                if (outcome?.kind === 'error') throw new Error(outcome.message);
                // Empty script: a benign terminal text response so loops don't hang.
                const message: ScriptedMessage = outcome?.message ?? { content: '' };
                const usage = outcome?.kind === 'message' ? outcome.usage : undefined;
                return { choices: [{ message }], usage: usage ?? null };
            },

            // Streaming seam: replays the SAME FIFO script as chunked deltas. Text is
            // split into a few pieces (so consumers see multiple `text` events); a
            // tool call's `arguments` is split across two chunks (so the agent's
            // accumulator is exercised); a final empty-delta chunk carries usage.
            createStream: (body: Record<string, unknown>): AsyncIterable<ChatChunk> => {
                this.record(body);
                const outcome = this.script.shift();
                const message: ScriptedMessage = outcome?.kind === 'message' ? outcome.message : { content: '' };
                const usage = outcome?.kind === 'message' ? outcome.usage : undefined;
                const error = outcome?.kind === 'error' ? outcome.message : undefined;

                async function* gen(): AsyncGenerator<ChatChunk> {
                    if (error) throw new Error(error);
                    // Text content → 2-3 deltas.
                    const content = message.content ?? '';
                    for (const piece of splitIntoChunks(content, 3)) {
                        if (piece) yield { choices: [{ delta: { content: piece } }] };
                    }
                    // Tool calls → opening chunk (id + name + first arg half), then a
                    // second chunk with the rest of the arguments. Exercises the
                    // index-keyed accumulator on the agent side.
                    for (const [index, tc] of (message.tool_calls ?? []).entries()) {
                        const args = tc.function.arguments ?? '';
                        const mid = Math.floor(args.length / 2);
                        yield {
                            choices: [{ delta: { tool_calls: [{ index, id: tc.id, function: { name: tc.function.name, arguments: args.slice(0, mid) } }] } }],
                        };
                        yield {
                            choices: [{ delta: { tool_calls: [{ index, function: { arguments: args.slice(mid) } }] } }],
                        };
                    }
                    // Final chunk carries usage (gateways send it on the last chunk).
                    yield { choices: [{ delta: {} }], usage: usage ? { prompt_tokens: usage.prompt_tokens ?? 0, completion_tokens: usage.completion_tokens ?? 0 } : null };
                }
                return gen();
            },
        },
    };
}

/** Split `s` into up to `n` roughly-equal non-empty pieces (≥2 when long enough). */
function splitIntoChunks(s: string, n: number): string[] {
    if (s.length === 0) return [];
    const parts = Math.min(n, Math.max(1, s.length));
    const size = Math.ceil(s.length / parts);
    const out: string[] = [];
    for (let i = 0; i < s.length; i += size) out.push(s.slice(i, i + size));
    return out;
}
